Add lock, unlock functionality

swayongshen opened the pull request:
Lock function adds a password to the state of the program and when the program exits, it uses this password to create a password protected zip file containing addressbook.json. The original addressbook.json is then deleted.

Unlock function removes the password from the state of the program. When the program exits, it creates an unprotected zip file containing addressbook.json. The original addressbook.json is then deleted.

swayongshen commented:
I  recommend you clone this https://github.com/swayongshen/tp/tree/add-lock and do some general testing.

NBH99 commented on src/main/java/seedu/address/logic/parser/LockCommandParser.java:
```diff
@@ -0,0 +1,26 @@
+        String trimmedArgs = args.trim();
+        String[] passwords = trimmedArgs.split("\\s+");
+
+        if (passwords.length < 1 || passwords.length > 2) {
```
The code here does not capture the error where user input "lock" has no password.

swayongshen replied:
Thanks for the feedback! Turns out String.split() will return an array with 1 element when the string is an empty string. A similar problem in UnlockCommandParser has also been fixed.

NBH99 reviewed (commented):
After locking the clientbook and exit, the program cannot run anymore.

jay9645 commented on docs/UserGuide.md:
```diff
@@ -154,6 +156,24 @@ Exits the program.
+
+Unlocks ClientBook.
+
+Format: `unlock [CURRENT_PASSWORD]`
```
I'm a little confused about why current_password is wrapped in square brackets. Does this mean that if a password has not been set, then unlock command does not need a password?

swayongshen replied:
Yes, the original intent is for unlock command to work even without CURRENT_PASSWORD if ClientBook is not locked. However, in hindsight this is confusing for the user and I have changed CURRENT_PASSWORD to be a compulsory argument in the UG.

jay9645 commented on src/main/java/seedu/address/logic/commands/LockCommand.java:
```diff
@@ -0,0 +1,70 @@
+        return new CommandResult(MESSAGE_LOCK_SUCCESS);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof LockCommand // instanceof handles nulls
+                && currentPassword.equals(((LockCommand) other).currentPassword)
+                && newPassword.equals(((LockCommand) other).newPassword)); // state check
+    }
```
For the return statement of this method, the comments helped me to understand the code easily which is already great, but I was thinking whether the expression would be a little less complicated if you separate the boolean statement into multiple named boolean variables.

For example
```java
    @Override
    public boolean equals(Object other) {
        isSameObject = other == this;
        if (isSameObject) {
            return true;
        } else {
            isLockCommandObj = other instanceof LockCommand; // instanceof handles nulls
            isCurrentPasswordEqual = currentPassword.equals(((LockCommand) other).currentPassword);
            isNewPasswordEqual = newPassword.equals(((LockCommand) other).newPassword);

            return isSameObject || (isLockCommandObj && isCurrentPasswordEqual && isNewPasswordEqual);
        }
    }
```

swayongshen replied:
Appreciate the feedback! However that was the format that came with the incumbent code e.g. DeleteCommand's equals().

jay9645 replied:
I see, then let's follow the format 😁

jay9645 commented on src/main/java/seedu/address/logic/commands/UnlockCommand.java:
```diff
@@ -0,0 +1,52 @@
+        if (!authentication.getPassword().equals(currentPassword)) {
+            throw new CommandException(MESSAGE_INCORRECT_PASSWORD);
+        }
+        authentication.removePassword();
```
If we remove the password here,  does that mean that every time the user unlocks the ClientBook, the next time he locks he needs to set a password again? 

If that's the case, then I'm thinking if it is possible that we don't remove password on unlock and store the password, so the user only needs to specify a password for the lock command when he/she first adds or changes his/her password. Then subsequently, if he/she locks and unlocks it will just use the same pasword.

edit: after thinking about this more, I realised that its not so easy to store the password securely within the app (I was thinking  in UserPrefs? but not sure if it is secure or even possible), and perhaps this one time lock and unlock is enough. What do you think?

swayongshen replied:
That is a great suggestion actually. 

One possible way would be to encrypt a password file using a secret key within our code which when the program runs, will be decrypted and stored in program state and encrypted immediately. 

The user can then enter lock without any parameters to lock using the old password.

However, I do feel that the effort to return ratio is rather minimal as it does not make our program any more tailored for our user than what already exists.

jay9645 replied:
I agree that the effort to reward ratio is pretty low since it's not a particularly core feature of our application. In that case, let's keep this extension in mind, and if we have time after other core features then we can possibly come back to it in the future.

jay9645 commented on src/main/java/seedu/address/storage/Authentication.java:
```diff
@@ -0,0 +1,202 @@
+        dataJson.delete();
+    }
+
+    private boolean attemptUnzip(String password) throws ZipException {
+        ZipFile dataZip = new ZipFile(this.getZipPath());
+        //First check if zip file encrypted and set the password to be used.
+        try {
+            if (dataZip.isEncrypted()) {
+                dataZip.setPassword(password.toCharArray());
+            }
+        } catch (ZipException e) {
+            logger.info("Error attempting to check zip file at: " + this.getZipPath());
+            throw e;
+        }
+
+        //Attempt to unzip the file using the provided password.
+        try {
+            dataZip.extractAll(this.getFolderPath());
+            return true;
+        } catch (ZipException e) {
+            logger.info("Error when unzipping file at: " + this.getZipPath());
+            return false;
+        }
+    }
```
This method is fine, but are the 2 try-catch blocks are mainly to differentiate between the 2 types of exceptions? I was thinking whether you can condense it into one try-catch block, and in the catch block, you differentiate the 2 cases using the ZipException that was caught.

Actually I was looking at the docs, and found that they didn't really specify what ZipException each method threw, but on their ZipException class, there is a getType() method, which returns a ZipException.Type. (https://javadoc.io/static/net.lingala.zip4j/zip4j/2.7.0/net/lingala/zip4j/exception/ZipException.Type.html) 

Perhaps you could check if it is possible to differentiate the cases by checking the exception type? If not then your method is completely fine as it is

swayongshen replied:
Good point there, I was looking at the same API a few days ago too. The methods ZipFile.isEncrypted() and ZipFile.extractAll() did not specify which type of ZipException was thrown. The try blocks are separated because I was using the second exception to return false, while the first try block was used to catch the error thrown when checking if the zip file is encrypted.

jay9645 replied:
I see, since the API doesn't specify then it's difficult to catch the different cases ourselves. In that case then two try-catch blocks is a nice way to circumvent that👍🏼

jay9645 commented on src/main/java/seedu/address/storage/Authentication.java:
```diff
@@ -0,0 +1,202 @@
+            logger.info("Failed to compress data .json file into a zip file");
+            throw e;
+        }
+
+
+
+
```
Just a little bit of unnecessary white space :)

swayongshen replied:
Fixed.

jay9645 commented on src/main/java/seedu/address/storage/Authentication.java:
```diff
@@ -0,0 +1,202 @@
+        }
+    }
+
+    private boolean attemptUnzip() throws ZipException {
+        return attemptUnzip("");
+    }
```
Is there some reason you chose to overload the attemptUnzip() method with this? It seems to me that you could directly call attemptUnzip("") where you want to call this method.

swayongshen replied:
It was to make it clear during usage that there is no password involved as someone reading it may interpret as entering an empty password.

jay9645 replied:
I see, that makes sense. Perhaps it could be clearer if you leave a comment or rename the method to indicate that it is for cases where no password was involved?

swayongshen replied:
I agree, I have updated the method's name to improve clarity.

jay9645 reviewed (commented):
Awesome code quality! Just a couple of things that I hope to clarify with the comments I left.

swayongshen commented:
I have implemented the new feature of saving the old password. When the user calls the lock command to set a new password, the password will be stored in an encrypted file named keystore in /data/. 

After unlocking ClientBook, the next time the user tries to lock ClientBook without a password, the old password will be decrypted from the password file and added to the state of the program and the password file will be encrypted immediately.

jay9645 commented:
Just did some testing, feature works perfectly! The only small issue is when you input `unlock` without any parameters, the feedback message to user is 
```
Invalid command format! 
lock: Locks ClientBook with a password.
Parameters: CURRENT_PASSWORD(if any) NEW_PASSWORD
Example: lock 12345
```

And also I think maybe it would make it clearer for the user if we add in the user guide a sentence to specify that when lock is used without parameters, then it will be locked with the current stored password.

After these 2 then its LGTM for me!
